libs/ledger-live-common/src/families/tron/signOperation.ts

```diff
@@ -0,0 +1,139 @@
+import { Observable } from "rxjs";
+import BigNumber from "bignumber.js";
+import { AccountBridge, TokenAccount } from "@ledgerhq/types-live";
+import { Transaction, TransactionStatus, TronAccount } from "./types";
+import { buildOptimisticOperation } from "./buildOptimisticOperation";
+import { TronSendTrc20ToNewAccountForbidden } from "./errors";
+import signTransaction from "../../hw/signTransaction";
+import { withDevice } from "../../hw/deviceAccess";
+import { getEstimatedFees } from "./logic";
+import {
+  claimRewardTronTransaction,
+  createTronTransaction,
+  fetchTronAccount,
+  fetchTronContract,
+  freezeTronTransaction,
+  legacyUnfreezeTronTransaction,
+  unDelegateResourceTransaction,
+  unfreezeTronTransaction,
+  voteTronSuperRepresentatives,
+  withdrawExpireUnfreezeTronTransaction,
+} from "./api";
+
+export const signOperation: AccountBridge<
+  Transaction,
+  TransactionStatus,
+  TronAccount
+>["signOperation"] = ({ account, transaction, deviceId }) =>
+  withDevice(deviceId)(
+    transport =>
+      new Observable(o => {
+        async function main() {
+          const subAccount = transaction.subAccountId
+            ? account.subAccounts?.find(sa => sa.id === transaction.subAccountId)
+            : undefined;
+          const isContractAddressRecipient = !!(await fetchTronContract(transaction.recipient));
+          const fee = await getEstimatedFees(account, transaction, isContractAddressRecipient);
+          const balance = subAccount
+            ? subAccount.balance
+            : BigNumber.max(0, account.spendableBalance.minus(fee));
+
+          if (transaction.useAllAmount) {
+            transaction = { ...transaction }; // transaction object must not be mutated
+            transaction.amount = balance; // force the amount to be the max
+          }
+
+          // send trc20 to a new account is forbidden by us (because it will not activate the account)
+          if (
+            transaction.recipient &&
+            transaction.mode === "send" &&
+            subAccount &&
+            subAccount.type === "TokenAccount" &&
+            subAccount.token.tokenType === "trc20" &&
+            !isContractAddressRecipient && // send trc20 to a smart contract is allowed
+            (await fetchTronAccount(transaction.recipient)).length === 0
+          ) {
+            throw new TronSendTrc20ToNewAccountForbidden();
+          }
```

Comment:
Looks like most of the "prepareTransaction" is actually done here. I tried to create the `prepareTransactionForSignature` to at least remove some of this logic, but it *really* feels wrong to have this here and nothing in `prepareTransaction`